homeassistant/components/usb/__init__.py

```diff
@@ -0,0 +1,138 @@
+"""The USB Discovery integration."""
+from __future__ import annotations
+
+import dataclasses
+import datetime
+import logging
+import sys
+
+from serial.tools.list_ports import comports
+from serial.tools.list_ports_common import ListPortInfo
+
+from homeassistant import config_entries
+from homeassistant.const import EVENT_HOMEASSISTANT_STARTED, EVENT_HOMEASSISTANT_STOP
+from homeassistant.core import Event, HomeAssistant, callback
+from homeassistant.helpers.event import async_track_time_interval
+from homeassistant.helpers.typing import ConfigType
+from homeassistant.loader import async_get_usb
+
+from .flow import FlowDispatcher, USBFlow
+from .models import USBDevice
+from .utils import usb_device_from_port
+
+_LOGGER = logging.getLogger(__name__)
+
+# Perodic scanning only happens on non-linux systems
+SCAN_INTERVAL = datetime.timedelta(minutes=60)
+
+
+async def async_setup(hass: HomeAssistant, config: ConfigType) -> bool:
+    """Set up the USB Discovery integration."""
+    usb = await async_get_usb(hass)
+    usb_discovery = USBDiscovery(hass, FlowDispatcher(hass), usb)
+    await usb_discovery.async_setup()
+    return True
+
+
+class USBDiscovery:
+    """Manage USB Discovery."""
+
+    def __init__(
+        self,
+        hass: HomeAssistant,
+        flow_dispatcher: FlowDispatcher,
+        usb: list[dict[str, str]],
+    ) -> None:
+        """Init USB Discovery."""
+        self.hass = hass
+        self.flow_dispatcher = flow_dispatcher
+        self.usb = usb
+        self.seen: set[tuple[str, ...]] = set()
+
+    async def async_setup(self) -> None:
+        """Set up USB Discovery."""
+        if not await self._async_start_monitor():
+            await self._async_start_scanner()
+        self.hass.bus.async_listen_once(EVENT_HOMEASSISTANT_STARTED, self.async_start)
+
+    async def async_start(self, event: Event) -> None:
+        """Start USB Discovery and run a manual scan."""
+        self.flow_dispatcher.async_start()
+        await self.hass.async_add_executor_job(self.scan_serial)
+
+    async def _async_start_scanner(self) -> None:
+        """Perodic scan with pyserial when the observer is not available."""
+        stop_track = async_track_time_interval(
+            self.hass, lambda now: self.scan_serial(), SCAN_INTERVAL
+        )
+        self.hass.bus.async_listen_once(
+            EVENT_HOMEASSISTANT_STOP, callback(lambda event: stop_track())
+        )
+
+    async def _async_start_monitor(self) -> bool:
+        """Start monitoring hardware with pyudev."""
+        if not sys.platform.startswith("linux"):
+            return False
+        from pyudev import (  # pylint: disable=import-outside-toplevel
+            Context,
+            Monitor,
+            MonitorObserver,
+        )
+
+        try:
+            context = Context()
```

Comment:
This throws an exception if libudev is not installed. Is libudev now a requirement for Core installations (i.e. a breaking change), or should the error be handled gracefully? Sorry to comment on a merged PR, but I wasn't sure if this is an issue or not.

```
2021-08-20 18:11:51 ERROR (MainThread) [homeassistant.setup] Error during setup of component usb
Traceback (most recent call last):
  File "/home/keith/src/core/homeassistant/setup.py", line 255, in _async_setup_component
    result = await task
  File "/home/keith/src/core/homeassistant/components/usb/__init__.py", line 33, in async_setup
    await usb_discovery.async_setup()
  File "/home/keith/src/core/homeassistant/components/usb/__init__.py", line 54, in async_setup
    if not await self._async_start_monitor():
  File "/home/keith/src/core/homeassistant/components/usb/__init__.py", line 83, in _async_start_monitor
    context = Context()
  File "/home/keith/src/core/venv/lib/python3.9/site-packages/pyudev/core.py", line 61, in __init__
    self._libudev = load_ctypes_library('udev', SIGNATURES, ERROR_CHECKERS)
  File "/home/keith/src/core/venv/lib/python3.9/site-packages/pyudev/_ctypeslib/utils.py", line 56, in load_ctypes_library
    lib = CDLL(library_name, use_errno=True)
  File "/nix/store/81lwy2hfqj4c1943b1x8a0qsivjhdhw9-python3-3.9.6/lib/python3.9/ctypes/__init__.py", line 374, in __init__
    self._handle = _dlopen(self._name, mode)
OSError: libudev.so.1: cannot open shared object file: No such file or directory
```

Reply:
We should trap the OSError and fallback

Reply:
https://github.com/home-assistant/core/pull/54967